model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaMapRoleEntityDelegate.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.storage.jpa.role;
+
+import org.keycloak.models.map.common.StringKeyConverter;
+import org.keycloak.models.map.role.MapRoleEntityDelegate;
+import org.keycloak.models.map.storage.jpa.role.delegate.JpaRoleDelegateProvider;
+import org.keycloak.models.map.storage.jpa.role.entity.JpaRoleCompositeEntity;
+import org.keycloak.models.map.storage.jpa.role.entity.JpaRoleCompositeEntityKey;
+import org.keycloak.models.map.storage.jpa.role.entity.JpaRoleEntity;
+
+import javax.persistence.EntityManager;
+import javax.persistence.Query;
+import javax.persistence.TypedQuery;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Delegate for the JPA implementation for MapRoleEntityDelegate.
+ * It will delegate all access to the composite roles to a separate table.
+ *
+ * For performance reasons, it caches the composite roles within the session if they have already been retrieved.
+ * This relies on the behavior of {@link org.keycloak.models.map.storage.jpa.JpaMapKeycloakTransaction} that
+ * each entity is created only once within each session.
+ *
+ * @author Alexander Schwartz
+ */
+public class JpaMapRoleEntityDelegate extends MapRoleEntityDelegate {
+    private final EntityManager em;
+
+    private Set<String> compositeRoles;
+
+    public JpaMapRoleEntityDelegate(JpaRoleEntity original, EntityManager em) {
+        super(new JpaRoleDelegateProvider(original, em));
+        this.em = em;
+    }
+
+    @Override
+    public Set<String> getCompositeRoles() {
```

Comment:
I had in mind something like this:
```
    // method is not used, it's overridden by JpaMapRoleEntityDelegate
    @Override
    public Set<String> getCompositeRoles() {
        return metadata.getCompositeRoles();
    }

    // method is not used, it's overridden by JpaMapRoleEntityDelegate
    @Override
    public void setCompositeRoles(Set<String> compositeRoles) {
        metadata.setCompositeRoles(compositeRoles);
    }

    // method is not used, it's overridden by JpaMapRoleEntityDelegate
    @Override
    public void addCompositeRole(String roleId) {
        metadata.addCompositeRole(roleId);
    }

    // method is not used, it's overridden by JpaMapRoleEntityDelegate
    @Override
    public void removeCompositeRole(String roleId) {
        metadata.removeCompositeRole(roleId);
    }
```

Reply:
I do, thank you ... this change makes it even visible to the caller, not just to a "code-reader" ... I'd only throw `UnsupportedOperationException` instead of `IllegalArgumentException` as it seems to me it describes the situation a little bit better. But feel free to leave it as `IllegalArgumentException`, it's more an optional suggestion.